# Introduction

The Quickomics tool can be accessed via the link [https://quickomics.bxgenomics.com](https://quickomics.bxgenomics.com){target="_blank"}. Implemented through R Shiny, it helps with visualizing statistical analysis results for RNAseq and Proteomics datasets. This supplemental tutorial provides a detailed guide on using the different functionalities and customizing the tools to best fit individual analysis and plotting needs, using published RNAseq dataset and proteomics dataset as examples ([Gyoneva et al., 2019](https://www.life-science-alliance.org/content/2/6/e201900453){target="_blank"}; [Connor-Robson et al., 2019](https://www.sciencedirect.com/science/article/pii/S0969996119300968?via%3Dihub){target="_blank"}).

[![](https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS1_1.png)](https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS1_1.png)

The interface contains multiple tabs, corresponding to different functional modules, that can be accessed on top panel of the webpage. Users have the option to upload their own dataset or choose from existing demo datasets for visualization.

## Integrated Gene Set Query

To supply genes in pre-defined gene sets to plotting functions in Quickomics, we developed xGenesets API for Quickomics to query and retrieve a list of genes/proteins when it is needed. The lists can be defined based on pathways, and gene sets from various sources, including KEGG Pathways (Kanehisa and Goto, 2000), WikiPathways (Martens et al., 2020), Small Molecule Pathway Database (Frolkis et al., 2010), Reactome (Jassal et al., 2020), Gene Ontology (The Gene Ontology Consortium, 2019), Molecular Signatures (Subramanian et al., 2005; Liberzon et al., 2015), and LIPID MAPS Proteome Database (Cotter et al., 2006). xGenesets database currently has 175,537 gene sets stored in MySQL tables, covering human (115,497 records), mouse (30,554 records) and rat (29,486 records) gene sets. xGenesets features personalized settings, dynamic drop-down list upon typing for quick gene set selection, full gene set database browsing, and a list of genes from any gene sets. Throughout Quickomics, users interact with this convenient tool in appropriate interface when a gene set is needed by checking “Geneset” option and then clicking on “Select Geneset” to pick a set from a table of available gene sets in a popup window as shown in the following screenshot.

[![](https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS1.1_1.png)](https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS1.1_1.png)

For developers of web-based applications (general user can skip this paragraph), xGenesets can be easily embedded in any modern websites and online applications, e.g., R Shiny web apps, by leveraging jQuery, DataTables, and bootstrap (either v3 or v4) JavaScript libraries. xGenesets API is freely available for all usages. Examples of using xGenesets API and guidance on setting up the API in your own applications can be found at [https://bxaf.net/genesets](https://bxaf.net/genesets){target="_blank"}.

## Launch project saved on server

In this updated version 2.0 of Quickomics, we added this new function to launch projects saved on the server directly using URL.

To view your own datasets, you can upload RData files using browser, but a better way is to put the RData files on your server where you host QuickOmics and access them directly using a URL. We suggest the following:

**1. Prepare three files for each project**, `projectID.RData` and `projectID_network.RData` file, as well as a `projectID.csv` file.

The csv file should have four columns: 

```
Name,ShortName,ProjectID,Species
```

Two optional columns for the csv file include:  `ExpressionUnit` (you can enter something like log2(TPM+0.1)),  and `Path` (optional path for raw or analyzed data, e.g. Z:/share/GSE158470/).

**2. Put the three files you prepared in a special folder**, and use URL to access them.

The simplest way is to put the three files in the unlisted/ folder under Quickomics folder. Then use a URL like [https://quickomics.bxgenomics.com/?unlisted=ProjectID](https://quickomics.bxgenomics.com/?unlisted=ProjectID){target="_blank"} (replace quickomics.bxgenomics.com with your own instance URL).

You can also create a config.csv file under Quickomics folder, for example:

```
category,value
server_dir, /hpc/data/server_files
test_dir,/hpc/data/test_fiels
```

Just make sure the two folders in server_dir and test_dir can be read by the Shiny App.

Then you can copy the three files prepared in step 1) into the server_dir, and use URL like [https://quickomics.bxgenomics.com/?serverfile=ProjectID](https://quickomics.bxgenomics.com/?serverfile=ProjectID){target="_blank"}
Or you can copy the three files prepared in step 1) into the test_dir, and use URL like [https://quickomics.bxgenomics.com/?testfile=ProjectID](https://quickomics.bxgenomics.com/?testfile=ProjectID){target="_blank"}
Using `config.csv` is preferred as the data are separated from the QuickOmics Shiny App folder.


## ShinyOne

In Quickomics v2.0, we also introduce ShinyOne, a web-based table listing existing projects loaded into Quickomics.

We have pre-loaded several demo data to Quickomics, and the ShinyOne demo site is:

[http://demo.bxgenomics.com:8185/](http://demo.bxgenomics.com:8185/){target="_blank"}
