<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 DEG Module | Supplementary Materials - Quickomics v2.0</title>
  <meta name="description" content="Chapter 5 DEG Module | Supplementary Materials - Quickomics v2.0" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 DEG Module | Supplementary Materials - Quickomics v2.0" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 DEG Module | Supplementary Materials - Quickomics v2.0" />
  
  
  



<meta name="date" content="2023-10-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="qc-plots-module.html"/>
<link rel="next" href="heatmap-module.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Quickomics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#integrated-gene-set-query"><i class="fa fa-check"></i><b>1.1</b> Integrated Gene Set Query</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#launch-project-saved-on-server"><i class="fa fa-check"></i><b>1.2</b> Launch project saved on server</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#shinyone"><i class="fa fa-check"></i><b>1.3</b> ShinyOne</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dataset-module.html"><a href="dataset-module.html"><i class="fa fa-check"></i><b>2</b> Dataset Module</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dataset-module.html"><a href="dataset-module.html#upload-files"><i class="fa fa-check"></i><b>2.1</b> Upload Files</a></li>
<li class="chapter" data-level="2.2" data-path="dataset-module.html"><a href="dataset-module.html#prepare-r-data-files-by-computational-biologists"><i class="fa fa-check"></i><b>2.2</b> Prepare R Data Files by Computational Biologists</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="dataset-module.html"><a href="dataset-module.html#example-r-script-to-prepare-r-data-files-from-rna-seq-results"><i class="fa fa-check"></i><b>2.2.1</b> Example R script to prepare R data files from RNA-Seq results</a></li>
<li class="chapter" data-level="2.2.2" data-path="dataset-module.html"><a href="dataset-module.html#example-r-script-to-prepare-r-data-files-from-proteomics-results"><i class="fa fa-check"></i><b>2.2.2</b> Example R script to prepare R data files from proteomics results</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="dataset-module.html"><a href="dataset-module.html#sample-table"><i class="fa fa-check"></i><b>2.3</b> Sample Table</a></li>
<li class="chapter" data-level="2.4" data-path="dataset-module.html"><a href="dataset-module.html#project-overview"><i class="fa fa-check"></i><b>2.4</b> Project Overview</a></li>
<li class="chapter" data-level="2.5" data-path="dataset-module.html"><a href="dataset-module.html#result-table"><i class="fa fa-check"></i><b>2.5</b> Result Table</a></li>
<li class="chapter" data-level="2.6" data-path="dataset-module.html"><a href="dataset-module.html#data-table"><i class="fa fa-check"></i><b>2.6</b> Data Table</a></li>
<li class="chapter" data-level="2.7" data-path="dataset-module.html"><a href="dataset-module.html#protein-gene-names"><i class="fa fa-check"></i><b>2.7</b> Protein Gene Names</a></li>
<li class="chapter" data-level="2.8" data-path="dataset-module.html"><a href="dataset-module.html#help"><i class="fa fa-check"></i><b>2.8</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="groups-and-samples-module.html"><a href="groups-and-samples-module.html"><i class="fa fa-check"></i><b>3</b> Groups and Samples Module</a>
<ul>
<li class="chapter" data-level="3.1" data-path="groups-and-samples-module.html"><a href="groups-and-samples-module.html#select-groups-and-samples"><i class="fa fa-check"></i><b>3.1</b> Select groups and samples</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="qc-plots-module.html"><a href="qc-plots-module.html"><i class="fa fa-check"></i><b>4</b> QC Plots Module</a>
<ul>
<li class="chapter" data-level="4.1" data-path="qc-plots-module.html"><a href="qc-plots-module.html#pca-plot"><i class="fa fa-check"></i><b>4.1</b> PCA Plot</a></li>
<li class="chapter" data-level="4.2" data-path="qc-plots-module.html"><a href="qc-plots-module.html#covariates"><i class="fa fa-check"></i><b>4.2</b> Covariates</a></li>
<li class="chapter" data-level="4.3" data-path="qc-plots-module.html"><a href="qc-plots-module.html#alignqc"><i class="fa fa-check"></i><b>4.3</b> AlignQC</a></li>
<li class="chapter" data-level="4.4" data-path="qc-plots-module.html"><a href="qc-plots-module.html#eigenvalues"><i class="fa fa-check"></i><b>4.4</b> Eigenvalues</a></li>
<li class="chapter" data-level="4.5" data-path="qc-plots-module.html"><a href="qc-plots-module.html#pca-3d-plot"><i class="fa fa-check"></i><b>4.5</b> PCA 3D Plot</a></li>
<li class="chapter" data-level="4.6" data-path="qc-plots-module.html"><a href="qc-plots-module.html#pca-3d-interactive"><i class="fa fa-check"></i><b>4.6</b> PCA 3D Interactive</a></li>
<li class="chapter" data-level="4.7" data-path="qc-plots-module.html"><a href="qc-plots-module.html#sample-sample-distance"><i class="fa fa-check"></i><b>4.7</b> Sample-sample Distance</a></li>
<li class="chapter" data-level="4.8" data-path="qc-plots-module.html"><a href="qc-plots-module.html#dendrograms"><i class="fa fa-check"></i><b>4.8</b> Dendrograms</a></li>
<li class="chapter" data-level="4.9" data-path="qc-plots-module.html"><a href="qc-plots-module.html#box-plot"><i class="fa fa-check"></i><b>4.9</b> Box Plot</a></li>
<li class="chapter" data-level="4.10" data-path="qc-plots-module.html"><a href="qc-plots-module.html#cv-distribution"><i class="fa fa-check"></i><b>4.10</b> CV Distribution</a></li>
<li class="chapter" data-level="4.11" data-path="qc-plots-module.html"><a href="qc-plots-module.html#help-1"><i class="fa fa-check"></i><b>4.11</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="deg-module.html"><a href="deg-module.html"><i class="fa fa-check"></i><b>5</b> DEG Module</a>
<ul>
<li class="chapter" data-level="5.1" data-path="deg-module.html"><a href="deg-module.html#deg-counts"><i class="fa fa-check"></i><b>5.1</b> DEG Counts</a></li>
<li class="chapter" data-level="5.2" data-path="deg-module.html"><a href="deg-module.html#volcano-plot-static"><i class="fa fa-check"></i><b>5.2</b> Volcano Plot (Static)</a></li>
<li class="chapter" data-level="5.3" data-path="deg-module.html"><a href="deg-module.html#volcano-plot-interactive"><i class="fa fa-check"></i><b>5.3</b> Volcano Plot (Interactive)</a></li>
<li class="chapter" data-level="5.4" data-path="deg-module.html"><a href="deg-module.html#degs-and-deps-in-two-comparisons"><i class="fa fa-check"></i><b>5.4</b> DEGs and DEPs in Two Comparisons</a></li>
<li class="chapter" data-level="5.5" data-path="deg-module.html"><a href="deg-module.html#data-table-1"><i class="fa fa-check"></i><b>5.5</b> Data Table</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="heatmap-module.html"><a href="heatmap-module.html"><i class="fa fa-check"></i><b>6</b> Heatmap Module</a>
<ul>
<li class="chapter" data-level="6.1" data-path="heatmap-module.html"><a href="heatmap-module.html#static-heatmap-layout-1"><i class="fa fa-check"></i><b>6.1</b> Static Heatmap Layout 1</a></li>
<li class="chapter" data-level="6.2" data-path="heatmap-module.html"><a href="heatmap-module.html#static-heatmap-layout-2"><i class="fa fa-check"></i><b>6.2</b> Static Heatmap Layout 2</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="expression-plot-module.html"><a href="expression-plot-module.html"><i class="fa fa-check"></i><b>7</b> Expression Plot Module</a>
<ul>
<li class="chapter" data-level="7.1" data-path="expression-plot-module.html"><a href="expression-plot-module.html#browsing"><i class="fa fa-check"></i><b>7.1</b> Browsing</a></li>
<li class="chapter" data-level="7.2" data-path="expression-plot-module.html"><a href="expression-plot-module.html#searched-expression-data"><i class="fa fa-check"></i><b>7.2</b> Searched Expression Data</a></li>
<li class="chapter" data-level="7.3" data-path="expression-plot-module.html"><a href="expression-plot-module.html#data-table-2"><i class="fa fa-check"></i><b>7.3</b> Data Table</a></li>
<li class="chapter" data-level="7.4" data-path="expression-plot-module.html"><a href="expression-plot-module.html#result-table-1"><i class="fa fa-check"></i><b>7.4</b> Result Table</a></li>
<li class="chapter" data-level="7.5" data-path="expression-plot-module.html"><a href="expression-plot-module.html#rank-abundance-curve"><i class="fa fa-check"></i><b>7.5</b> Rank Abundance Curve</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gene-set-enrichment-module.html"><a href="gene-set-enrichment-module.html"><i class="fa fa-check"></i><b>8</b> Gene Set Enrichment Module</a>
<ul>
<li class="chapter" data-level="8.1" data-path="gene-set-enrichment-module.html"><a href="gene-set-enrichment-module.html#gene-set-enrichment"><i class="fa fa-check"></i><b>8.1</b> Gene Set Enrichment</a></li>
<li class="chapter" data-level="8.2" data-path="gene-set-enrichment-module.html"><a href="gene-set-enrichment-module.html#gene-expression"><i class="fa fa-check"></i><b>8.2</b> Gene Expression</a></li>
<li class="chapter" data-level="8.3" data-path="gene-set-enrichment-module.html"><a href="gene-set-enrichment-module.html#gene-set-heat-map"><i class="fa fa-check"></i><b>8.3</b> Gene Set Heat Map</a></li>
<li class="chapter" data-level="8.4" data-path="gene-set-enrichment-module.html"><a href="gene-set-enrichment-module.html#kegg-pathway-view"><i class="fa fa-check"></i><b>8.4</b> KEGG Pathway View</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="pattern-clustering-module.html"><a href="pattern-clustering-module.html"><i class="fa fa-check"></i><b>9</b> Pattern Clustering Module</a>
<ul>
<li class="chapter" data-level="9.1" data-path="pattern-clustering-module.html"><a href="pattern-clustering-module.html#clustering-of-centroid-profiles"><i class="fa fa-check"></i><b>9.1</b> Clustering of Centroid Profiles</a></li>
<li class="chapter" data-level="9.2" data-path="pattern-clustering-module.html"><a href="pattern-clustering-module.html#data-table-3"><i class="fa fa-check"></i><b>9.2</b> Data Table</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="correlation-network-module.html"><a href="correlation-network-module.html"><i class="fa fa-check"></i><b>10</b> Correlation Network Module</a>
<ul>
<li class="chapter" data-level="10.1" data-path="correlation-network-module.html"><a href="correlation-network-module.html#visnetwork"><i class="fa fa-check"></i><b>10.1</b> visNetwork</a></li>
<li class="chapter" data-level="10.2" data-path="correlation-network-module.html"><a href="correlation-network-module.html#data-table-4"><i class="fa fa-check"></i><b>10.2</b> Data Table</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="venn-diagram-module.html"><a href="venn-diagram-module.html"><i class="fa fa-check"></i><b>11</b> Venn Diagram Module</a>
<ul>
<li class="chapter" data-level="11.1" data-path="venn-diagram-module.html"><a href="venn-diagram-module.html#venn-diagram"><i class="fa fa-check"></i><b>11.1</b> Venn Diagram</a></li>
<li class="chapter" data-level="11.2" data-path="venn-diagram-module.html"><a href="venn-diagram-module.html#venn-diagram-black-white"><i class="fa fa-check"></i><b>11.2</b> Venn Diagram (black &amp; white)</a></li>
<li class="chapter" data-level="11.3" data-path="venn-diagram-module.html"><a href="venn-diagram-module.html#intersection-output"><i class="fa fa-check"></i><b>11.3</b> Intersection Output</a></li>
<li class="chapter" data-level="11.4" data-path="venn-diagram-module.html"><a href="venn-diagram-module.html#deg-table"><i class="fa fa-check"></i><b>11.4</b> DEG Table</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="output-module.html"><a href="output-module.html"><i class="fa fa-check"></i><b>12</b> Output Module</a></li>
<li class="chapter" data-level="13" data-path="qa.html"><a href="qa.html"><i class="fa fa-check"></i><b>13</b> Q&amp;A</a></li>
<li class="chapter" data-level="14" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>14</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Supplementary Materials - Quickomics v2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="deg-module" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> DEG Module<a href="deg-module.html#deg-module" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This section is designed to explore the differentially expressed genes (DEGs) and differentially expressed proteins (DEPs) in the dataset. The comparisons are defined by the User when they input the dataset. Please refer to our GitHub for more details on uploading datasets.</p>
<p>The “Volcano Plot (Static)” described in Section 5.2 and “Volcano Plot (Interactive)” described in Section 5.3 are ways to visualize each pair-wise comparison one by one. “DEGs/DEPs in Two Comparisons” described in Section 5.4 is a novel way to visualize the similarities and differences in two comparisons. Finally, “Data Table” described in Section 5.5 displays a searchable table containing all genes and their fold changes and P values in each comparison.</p>
<div id="deg-counts" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> DEG Counts<a href="deg-module.html#deg-counts" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>DEG Counts tab provides a summary of the number of significanly changed genes for each comparison, separating Up- and Down-regulated genes.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.1_1.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.1_1.png" /></a></p>
</div>
<div id="volcano-plot-static" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Volcano Plot (Static)<a href="deg-module.html#volcano-plot-static" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This plot reveals the differentially expressed genes provided in the input data file. By default, the Fold change Cutoff is set to 1.2 and the P value Cutoff is set to 0.01. Users have the option to change any of these parameters. For more advanced display options, they can set the “Show More Options” to “Yes” to customize other settings.</p>
<p>The red dots are indicative of all genes that pass the Fold Change and P value cutoff, while the green dots are indicative of genes that only pass the P value cutoff. Fifty random genes are labelled for both the up-regulated and down-regulated genes by default.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_1.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_1.png" /></a></p>
<p>For demonstration purposes, we have changed the following attributes to highlight a few genes in this dataset. We have selected the 2mon_KOvs2mon_WT comparison, where <a href="https://www.life-science-alliance.org/content/2/6/e201900453" target="_blank">Gyoneva et al., 2019</a> report immune response genes as the top differentially expressed category. To highlight these genes, in the example below, we select an “Immune Response” dataset from the Molecular Signatures Database (MSigDB) to highlight the genes by adding labels to the volcano plot. As MSigDB contains only human gene lists and the demo dataset is from mouse, Quickomics does a quick conversion of the names of Human genes to Mouse genes by changing the letters after the first one to lowercase. Additionally, in this example we removed some of the genes from the list to reduce size. This visualization was not available in the original publication, but clearly supports the finding that the KO genotypes altered the expression of immune-related genes.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_2.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_2.png" /></a></p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_3.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.2_3.png" /></a></p>
</div>
<div id="volcano-plot-interactive" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Volcano Plot (Interactive)<a href="deg-module.html#volcano-plot-interactive" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This next sub-tab is a very similar visualization as the previous one (5.2 Volcano Plot (Static)) with an additional capability where users can hover over a particular dot/gene on the plot to see more details. This plot has a few attributes that can be changed, and in the example below we have used the same cutoffs and values as in section 5.2.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.3_1.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.3_1.png" /></a></p>
</div>
<div id="degs-and-deps-in-two-comparisons" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> DEGs and DEPs in Two Comparisons<a href="deg-module.html#degs-and-deps-in-two-comparisons" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This next tab helps with identifying DEGs/DEPs that follow a similar trend in two comparisons. <a href="https://www.life-science-alliance.org/content/2/6/e201900453" target="_blank">Gyoneva et al., 2019</a>, the source of this demo dataset, report that the genes differentially expressed in <strong>Cx3cr1 KO</strong> microglia in the <strong>2mon KO</strong> are quite different from WT, while the differences from WT are smaller for the 1yr KO and 2yr KO, and the two older ages are similar to each other. We have plotted out the following comparisons that illustrate this conclusion.</p>
<p><strong>2mo_KOvs2mo_WT</strong> vs <strong>1yr_KOvs1yr_WT</strong> has a low correlation score. There are many genes that are colored yellow, meaning those genes were significant in <strong>2mo_KOvs2mo_WT (Y axis)</strong> but not in <strong>1yr_KOvs1yr_WT (X axis)</strong>.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_1.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_1.png" /></a></p>
<p>Similarly, we changed the X axis to <strong>2yr_KOvs2yr_WT</strong> to compare to <strong>2mo_KOvs2mo_WT</strong>, and it shows a very similar plot with a low correlation score of 0.178</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_2.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_2.png" /></a></p>
<p>Next, we changed the Y axis comparison to <strong>1yr_KOvs1yr_WT</strong> comparing it to <strong>2yr_KOvs2yr_WT</strong> on the X axis. This plot has a higher correlation score of 0.371 than the previous one, suggesting that these two comparisons are more similar. These comparisons represent additional insights supported by statistical analyses that were gained from the data and were not available in the original manuscript by Gyoneva et al.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_3.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.4_3.png" /></a></p>
</div>
<div id="data-table-1" class="section level2 hasAnchor" number="5.5">
<h2><span class="header-section-number">5.5</span> Data Table<a href="deg-module.html#data-table-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This tab lets the users view the DEGs/DEPs in a tabular format with a searchable feature. Users also have the ability to download the gene list as a CSV.</p>
<p><a href="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.5_1.png"><img src="https://interactivereport.github.io/Quickomics/tutorial/Figures/FigureS5.5_1.png" /></a></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="qc-plots-module.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="heatmap-module.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/Quickomics/edit/gh-pages/05-DEGModule.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/Quickomics/blob/gh-pages/05-DEGModule.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
